service-base/src/main/java/org/eclipse/hono/service/registration/BaseDeviceRegistryServer.java

```diff
@@ -0,0 +1,134 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+
+import java.util.Objects;
+import java.util.UUID;
+
+
+/**
+ * A server that implements Hono's <a href="https://www.eclipse.org/hono/api/Device-Registration-API/">Device Registration API</a> and
+ * <a href="https://www.eclipse.org/hono/api/Credentials-API/">Credentials API</a> as the default implementation of Hono's device registry.
+ *
+ * TODO: add metrics.
+ * TODO: refactor this into a more generic server class (eliminate device registry specific parts).
+ */
+public class BaseDeviceRegistryServer<T extends ServiceConfigProperties> extends AmqpServiceBase<T> {
+    private ConnectionFactory authenticationService;
+
+    /**
+     * Sets the factory to use for creating an AMQP 1.0 connection to
+     * the Authentication service.
+     *
+     * @param factory The factory.
+     * @throws NullPointerException if factory is {@code null}.
+     */
+    @Autowired
+    @Qualifier(AuthenticationConstants.QUALIFIER_AUTHENTICATION)
+    public void setAuthenticationServiceConnectionFactory(final ConnectionFactory factory) {
+        authenticationService = Objects.requireNonNull(factory);
+    }
+
+    @Override
+    protected void onRemoteConnectionOpen(final ProtonConnection connection) {
+        connection.setContainer(String.format("Hono-DeviceRegistry-%s:%d", getBindAddress(), getPort()));
+        setRemoteConnectionOpenHandler(connection);
+    }
+
+    @Override
+    protected void onRemoteConnectionOpenInsecurePort(final ProtonConnection connection) {
+        connection.setContainer(String.format("Hono-DeviceRegistry-%s:%d", getInsecurePortBindAddress(), getInsecurePort()));
+        setRemoteConnectionOpenHandler(connection);
+    }
+
+    private void setRemoteConnectionOpenHandler(final ProtonConnection connection) {
+        connection.sessionOpenHandler(remoteOpenSession -> handleSessionOpen(connection, remoteOpenSession));
+        connection.receiverOpenHandler(remoteOpenReceiver -> handleReceiverOpen(connection, remoteOpenReceiver));
+        connection.senderOpenHandler(remoteOpenSender -> handleSenderOpen(connection, remoteOpenSender));
+        connection.disconnectHandler(this::handleRemoteDisconnect);
+        connection.closeHandler(remoteClose -> handleRemoteConnectionClose(connection, remoteClose));
+        connection.openHandler(remoteOpen -> {
+            LOG.info("client [container: {}, user: {}] connected", connection.getRemoteContainer(), Constants.getClientPrincipal(connection).getName());
+            connection.open();
+            // attach an ID so that we can later inform downstream components when connection is closed
+            connection.attachments().set(Constants.KEY_CONNECTION_ID, String.class, UUID.randomUUID().toString());
+        });
+    }
+
+    private void handleSessionOpen(final ProtonConnection con, final ProtonSession session) {
```

Comment:
This is a good point. Most of the code in ´BaseDeviceRegistryServer´ is not specific for a device registry. The idea here was to keep the changes small in this step. I think @sysexcontrol plans to consolidate AMQP server code.